drivers/can/can_native_linux.c

```diff
@@ -0,0 +1,434 @@
+/*
+ * Copyright (c) 2022 Martin Jäger <martin@libre.solar>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_native_linux_can
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/drivers/can.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/net_pkt.h>
+#include <zephyr/net/socketcan.h>
+#include <zephyr/net/socketcan_utils.h>
+
+#include "can_utils.h"
+#include "can_native_linux_priv.h"
+
+LOG_MODULE_REGISTER(can_native_linux, CONFIG_CAN_LOG_LEVEL);
+
+struct can_filter_context {
+	can_rx_callback_t rx_cb;
+	void *cb_arg;
+	struct can_filter filter;
+};
+
+struct can_native_linux_data {
+	/* Linux socket file descriptor */
+	int dev_fd;
+
+	struct can_filter_context filters[CONFIG_CAN_MAX_FILTER];
+
+	bool loopback;
+	bool mode_fd;
+
+	struct k_mutex filter_mutex;
+	struct k_sem tx_sem;
+
+	struct k_thread rx_thread;
+
+	K_KERNEL_STACK_MEMBER(rx_thread_stack, CONFIG_CAN_NATIVE_LINUX_RX_THREAD_STACK_SIZE);
+};
+
+struct can_native_linux_config {
+	const char *if_name;
+};
+
+static void dispatch_frame(const struct device *dev, struct can_frame *frame)
+{
+	struct can_native_linux_data *data = dev->data;
+	can_rx_callback_t callback;
+	struct can_frame tmp_frame;
+
+	k_mutex_lock(&data->filter_mutex, K_FOREVER);
+
+	for (int filter_id = 0; filter_id < ARRAY_SIZE(data->filters); filter_id++) {
+		if (data->filters[filter_id].rx_cb == NULL) {
+			continue;
+		}
+
+		if (!can_utils_filter_match(frame,
+					    &data->filters[filter_id].filter)) {
+			continue;
+		}
+
+		/* Make a temporary copy in case the user modifies the message */
+		tmp_frame = *frame;
+
+		callback = data->filters[filter_id].rx_cb;
+		callback(dev, &tmp_frame, data->filters[filter_id].cb_arg);
+	}
+
+	k_mutex_unlock(&data->filter_mutex);
+}
+
+static void rx_thread(void *arg1, void *arg2, void *arg3)
+{
+	const struct device *dev = arg1;
+	struct can_native_linux_data *data = dev->data;
+	struct socketcan_frame sframe;
+	struct can_frame frame;
+	bool msg_confirm;
+	int count;
+
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("Starting Linux SocketCAN RX thread");
+
+	while (true) {
+		while (can_native_linux_wait_data(data->dev_fd) == 0) {
+			count = can_native_linux_read_data(data->dev_fd, (void *)(&sframe),
+							   sizeof(sframe), &msg_confirm);
+			if (msg_confirm) {
+				k_sem_give(&data->tx_sem);
+				if (!data->loopback) {
+					continue;
+				}
+			}
+			if (count <= 0) {
+				break;
+			}
+
+			socketcan_to_can_frame(&sframe, &frame);
+
+			LOG_DBG("Received %d bytes. Id: 0x%x, ID type: %s %s",
+				frame.dlc, frame.id,
+				frame.id_type == CAN_STANDARD_IDENTIFIER ?
+						"standard" : "extended",
+				frame.rtr == CAN_DATAFRAME ? "" : ", RTR frame");
+
+			dispatch_frame(dev, &frame);
+		}
+
+		/* short sleep required to avoid blocking the native thread */
```

Comment:
About the comment: Note that this is a Zephyr thread like any other

Reply:
IIRC, the main thread was also blocked when I didn't put the `k_sleep` here. Also the [documentation about limitations in Native POSIX](https://docs.zephyrproject.org/latest/boards/posix/native_posix/doc/index.html#important-limitations) states that busy wait is not allowed. So I suspected that the whole application runs in one Linux (native) thread and doesn't have the kind of scheduler we have on actual hardware. But my understanding may be completely wrong. I don't understand much about the details of `native_posix` implementation.

Reply:
>I suspected that the whole application runs in one Linux (native) thread

In native_posix each Zephyr thread is mapped to a separate linux pthread, and the scheduling between those threads is handled by the Zephyr kernel. The POSIX arch simulates/pretends that there is only a single core CPU, and only allows one of those Zephyr threads to run at a time as commanded by the Zephyr kernel. Apart from those threads there is an additional thread which handles the initial bring up (before the Zephyr kernel is started) and the main "HW loop" while the simulated CPU is sleeping. But for simplicity this "HW model" thread is also paused by any SW thread is running and viceversa.

A description like: `/* short sleep required to avoid blocking the whole native_posix process */` would be ok.

Reply:
Excellent, thanks for the explanation. Will update the comment accordingly.